File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prefix = [1 for _ in range(len(nums))]
        suffix = [1 for _ in range(len(nums))]
        ans = []

        for i in range(1, len(nums)):
            prefix[i] = nums[i - 1] * prefix[i - 1]
        for i in range(len(nums) - 2, -1, -1):
            suffix[i] = nums[i + 1] * suffix[i + 1]
        for i in range(len(nums)):
            ans.append(prefix[i]*suffix[i])
        return ans

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-1.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int l = 0, r = nums.size() - 1;
        while(l <= r){
            int mid = l + (r - l) / 2;
            if(nums[mid] == target) return mid;
            else if(nums[mid] < target){
                if(nums[nums.size()-1] >= target){
                    l = mid + 1;
                }
                else{
                    mid--;
                    while(mid >= 0){
                        if(nums[mid] == target) return mid;
                        mid--;
                    }
                    return -1;
                }
            }
            else{
                if(nums[0] <= target){
                    r = mid - 1;
                }
                else{
                    mid++;
                    while(mid < nums.size()){
                        if(nums[mid] == target) return mid;
                        mid++;
                    }
                    return -1;
                }
            }
        }
        return -1;
    }
};
